File: Infer.hs
-- {-# LANGUAGE ScopedTypeVariables #-}
import Text.Printf

-- enumerate :: (Num index, Enum index) => [item] -> [(index, item)]
-- enumerate :: [item] -> [(Int, item)]
enumerate = zip [(0 :: Int)..]

-- main :: IO ()
main = do
    let colors = ["red", "yellow", "blue"]
        indexedColors = enumerate colors  -- :: [(Int, String)]
        showColor = uncurry $ printf "Color %d: %s"
    mapM_ putStrLn $ map showColor indexedColors

File: infer.c
#include <stdio.h>

int main() {
  const char* colors[] = {"red", "yellow", "blue"};
  for (
    int index = 0;
    index < sizeof(colors) / sizeof(*colors);
    index += 1
  ) {
    printf("Color %d: %s\n", index, colors[index]);
  }
}
